### PR TITLE
Refactor the generation of the metadata for linking

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -256,22 +256,22 @@ pub fn each_linked_rlib(
...
-            LibSource::MetadataOnly => {
-                return Err(format!(
-                    "could not find rlib for: `{}`, found rmeta (metadata) file",
-                    name
-                ));
-            }
-            LibSource::None => return Err(format!("could not find rlib for: `{}`", name)),
+        let path = if let Some((path, _)) = &info.used_crate_source[&cnum].rlib {
+            path
+        } else if info.used_crate_source[&cnum].rmeta.is_some() {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:268]: The `info.used_crate_source[&cnum]` search can be done once instead of twice here.

**Comment by petrochenkov** [comment]
All changes look good to me, but the source of systematic regressions in https://github.com/rust-lang/rust/pull/86105#issuecomment-856243484 needs to be figured out.

You can break this PR into parts for that and assign them to me, I'll r+ them quickly if they don't cause perf regressions.

**Comment by bjorn3** [comment]
Rebased on top of  #86877 and removed 
279acba as I committed it directly to cg_clif's repo in a slightly changed form: https://github.com/bjorn3/rustc_codegen_cranelift/commit/fed71e344818dd2b1f2e48005b6d3da21cda30f5

**Comment by bjorn3** [comment]
I will do a final perf run on this PR. If it is a regression, I will close it.

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
Perf is much better this time! There are still regressions, but they are smaller and there are a couple of improvements up to 0.7%.

File: compiler/rustc_middle/src/middle/cstore.rs [hunk 2]
```diff
@@ -204,23 +204,18 @@ pub type CrateStoreDyn = dyn CrateStore + sync::Sync;
...
-// positions.
+// In order to get this left-to-right dependency ordering, we use the reverse
+// postorder of all crates putting the leaves at the right-most positions.
 pub fn used_crates(tcx: TyCtxt<'_>) -> Vec<CrateNum> {
-    let mut libs = tcx
-        .crates(())
+    tcx.postorder_cnums(())
         .iter()
+        .rev()
         .cloned()
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/middle/cstore.rs:213]:
```suggestion
        .copied()
```

File: compiler/rustc_middle/src/middle/cstore.rs [hunk 3]
```diff
@@ -204,23 +204,18 @@ pub type CrateStoreDyn = dyn CrateStore + sync::Sync;
...
                 return None;
             }
             Some(cnum)
         })
-        .collect::<Vec<_>>();
-    let mut ordering = tcx.postorder_cnums(()).to_owned();
-    ordering.reverse();
-    libs.sort_by_cached_key(|&a| ordering.iter().position(|x| *x == a));
-    libs
+        .collect::<Vec<_>>()
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/middle/cstore.rs:220]:
```suggestion
        .collect()
```

**Comment by petrochenkov** [comment]
I have no idea why this caused performance differences.
The produced result should be the same, and from code changes it also should be an optimization.

**Comment by petrochenkov** [comment]
r=me with the nits addressed.

**Comment by bjorn3** [comment]
Fixed the nits and also switched from `filter_map` to plain `filter` as nothing was mapped anyway.

**Comment by petrochenkov** [comment]
The function can probably be inlined at this point, it's used only once anyway.

**Comment by petrochenkov** [comment]
r=me with squashed commits and green CI.

**Comment by bjorn3** [comment]
Splat

@bors r=petrochenkov

(CI was green when I squashed. I didn't rebase, so the tree is identical.)

**Comment by bjorn3** [comment]
The dist-aarch64-apple job seems to not have started. This is the full log:

```
2021-07-07T13:35:42.5693605Z Found online and idle hosted runner(s) in the current repository's organization account that matches the required labels: 'macos-latest'
2021-07-07T13:35:42.5693641Z Waiting for a hosted runner in 'organization' to pick this job...
```

@bors retry

**Pull request merged**